### PR TITLE
Add shuffle action.

### PR DESCRIPTION
**Pull request by FloopCZ**
In my opinion, since there is a sort action, there should also be a shuffle action. Requesting review.

### CODE REVIEW

**Comment by rhalbersma** [comment]
I wonder [where you got this idea](http://stackoverflow.com/questions/30425417/how-to-write-a-range-v3-action-for-random-shuffle#comment48946078_30426249) ;) 

BTW, if this PR gets accepted, you probably should also systematically replace random_shuffle from the various unit tests (dogfooding and all that...).

File: include/range/v3/action/shuffle.hpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        /// @{
+        namespace action
+        {
+            struct shuffle_fn
+            {
+            private:
+                friend action_access;
+                template<typename Gen>
+                static auto bind(shuffle_fn shuffle, Gen && gen)
+                RANGES_DECLTYPE_AUTO_RETURN
```
Comment (ericniebler) [inline include/range/v3/action/shuffle.hpp]: This `bind` function should be constrained with `UniformRandomNumberGenerator<Gen>`.

**Comment by ericniebler** [comment]
That's the only nit. The rest looks good.

**Comment by FloopCZ** [comment]
Thank you for your feedback, it really should be constrained. Fixed and commited.

**Pull request merged**